app/src/main/java/org/schabi/newpipe/player/BasePlayer.java

```diff
@@ -1285,4 +1301,44 @@ private boolean isPlaybackResumeEnabled() {
         return prefs.getBoolean(context.getString(R.string.enable_watch_history_key), true)
                 && prefs.getBoolean(context.getString(R.string.enable_playback_resume_key), true);
     }
+
+    public void setTimer(int hourOfDay, int minute, Context ctx) {
+        long time = ((hourOfDay * 60) + minute) * 60 * 1000;
+        if (time == 0) {
+            Toast.makeText(ctx, ctx.getString(R.string.timer_select_time_message), Toast.LENGTH_SHORT).show();
+            return;
+        }
+        timer.cancel();
+        timer.purge();
+        timerHour = hourOfDay;
+        timerMinute = minute;
+        timer = new Timer();
+        timer.schedule(new TimerTask() {
+            @Override
+            public void run() {
+                new Handler(Looper.getMainLooper()).post(() -> {
+                    if (DEBUG) Log.d(TAG, "Timer finished: Stopping the player");
+                    onPause();
+                });
+
+            }
+        }, time);
+        Date d = new Date();
+        d.setTime(System.currentTimeMillis() + time);
+        Toast.makeText(ctx.getApplicationContext(), String.format(ctx.getString(R.string.player_stop_message), d.toString()), Toast.LENGTH_LONG).show();
```

Comment:
Format suggestion:
```java
        long time = ((hourOfDay * 60) + minute) * 60 * 1000;
        if (time == 0) {
            Toast.makeText(ctx, ctx.getString(R.string.timer_select_time_message), Toast.LENGTH_SHORT).show();
            return;
        }

        timer.cancel();
        timer.purge();
        timer = new Timer();
        timer.schedule(() -> {
            if (DEBUG) Log.d(TAG, "Timer finished: stopping the player");
            onPause();
        }, time);

        Date d = new Date();
        d.setTime(System.currentTimeMillis() + time);
        Toast.makeText(ctx.getApplicationContext(), String.format(ctx.getString(R.string.player_stop_message), d.toString()), Toast.LENGTH_LONG).show();
```

There is an issue: the user should be able to cancel the timer at any time, not just during the `Toast.LENGTH_LONG` time.

Reply:
Correct, we need a custom TimePickerDialog which allows us to add
1. A title which says something like "Select time to stop playback at"
2. Buttons at the bottom to stop the timer